packages/yarnpkg-pnp/sources/esm-loader/loader.ts

```diff
@@ -136,7 +153,41 @@ export async function getSource(urlString: string, context: any, defaultGetSourc
   const url = new URL(urlString);
   if (url.protocol !== `file:`) return defaultGetSource(url, context, defaultGetSource);
 
+  urlString = fileURLToPath(urlString);
+
+  if (realModules.has(urlString)) {
+    return {
+      source: await readFile(urlString),
+    };
+  }
+
+  const fakeModulePath = path.join(path.dirname(urlString), `noop.js`);
+
+  const require = createRequire(fakeModulePath);
+  const dynModule = require(urlString);
+
+  let exportStrings: Array<string> = [];
+  if (dynModule.__esModule === true) {
+    exportStrings = Object.getOwnPropertyNames(dynModule).map(propKey => {
+      if (propKey === `default`) {
+        return `export default cjs['default']`;
+      } else {
+        return `const __${propKey} = cjs['${propKey}'];\n export { __${propKey} as ${propKey} }`;
+      }
+    });
```

Comment:
Note that this has a few key differences from the Node.js implementation:
- It exposes all the properties of `exports`, while Node.js only exposes the one that are statically analyzable
- It evaluates CJS modules earlier: during the "resolve" phase rather than during the "evaluation" phase

While this works in the Babel repo, it has some subtle differences that would be _really hard_ to debug when they cause problems. We should either use `cjs-module-lexer` to detect the exports, or we should push for an official API in Node.js.

Reply:
Made it use `cjs-module-lexer` instead, it has some interesting results on [prettier](https://cdn.jsdelivr.net/npm/prettier@2.2.1/index.js) but otherwise it's looking good